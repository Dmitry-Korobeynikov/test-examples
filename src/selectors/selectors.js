import { createSelector } from 'reselect';

// simple selector
export const getField = (state) => {
  return state.field;
};

// composite selector
export const isFalsyField = createSelector(
  [getField],
  (field) => {
    return !field;
  });
