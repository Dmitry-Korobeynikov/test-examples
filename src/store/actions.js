import { SIMPLE_ACTION } from '../consts';
import { asyncLogic } from '../modules/logic';

export const simpleAction = (payload) => {
  return {
    type: SIMPLE_ACTION,
    payload,
  };
};

export const thunkAction = () => {
  return (dispatch, getState) => {
    const { field } = getState();

    if (typeof field === 'string') {
      dispatch(simpleAction(`not ${field}`));
    } else if (typeof field === 'boolean') {
      dispatch(simpleAction(!field));
    }
  };
};

export const thunkAsyncAction = () => {
  return (dispatch) => {
    dispatch(exports.thunkAction());

    asyncLogic()
      .then((result) => {
        dispatch(exports.simpleAction(result));
      });
  };
};
