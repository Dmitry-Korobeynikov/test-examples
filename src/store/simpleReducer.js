import { SIMPLE_ACTION } from '../consts';

export const initialState = {
  field: 'value',
};

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case SIMPLE_ACTION:
      return {
        ...state,
        field: action.payload,
      };
    default:
      return state;
  }
}
