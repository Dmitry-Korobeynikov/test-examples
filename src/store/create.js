import { createStore as _createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import simpleReducer from './simpleReducer';

const create = (initialState) => {
  const middlewares = [thunk];
  const finalCreateStore = applyMiddleware(...middlewares)(_createStore);

  return finalCreateStore(simpleReducer, initialState);
};

export const createStoreForTests = create;
export default create;
