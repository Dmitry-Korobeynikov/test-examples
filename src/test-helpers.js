import React from 'react';
import PropTypes from 'prop-types';
import { shallow } from 'enzyme';
import { createStoreForTests } from './store/create';

export const renderShallow = (Component, props = {}, initialState = undefined) => {
  const store = createStoreForTests(initialState);
  const context = { store };
  const childContextTypes = { store: PropTypes.shape({}) };

  return shallow(<Component {...props} />, { context, childContextTypes });
};

/*
 * shallowWrapper: it should be an object returned from enzyme.shallow(...)
 * childSelector: could be a component display name, component constructor, an object with props...
 *   (https://airbnb.io/enzyme/docs/api/selector.html)
 * propName: it's property name which used as a render prop
 */
export const callRenderProp = (shallowWrapper, childSelector, propName = 'children') => {
  return (...args) => {
    return shallowWrapper.find(childSelector).renderProp(propName)(...args);
  };
};
