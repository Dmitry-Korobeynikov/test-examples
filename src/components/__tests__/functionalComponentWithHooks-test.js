import { renderShallow } from '../../test-helpers';
import { getField } from '../../selectors/selectors';
import { simpleAction } from '../../store/actions';
import FunctionalComponentWithHooks from '../functionalComponentWithHooks';

jest.mock('react-redux', () => {
  return {
    // save other "react-redux" functions, for example it may be usefull
    // for components with "connect" to prevent tests from falling 
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn((mockSelector) => {
      return mockSelector();
    }),
    useDispatch: jest.fn(() => {
      return (mockAction) => {
        return mockAction();
      };
    }),
  };
});

jest.mock('react', () => {
  return {
    ...jest.requireActual('react'),
    useCallback: jest.fn((callback) => {
      return callback;
    }),
  };
});

jest.mock('../../selectors/selectors');
jest.mock('../../store/actions', () => {
  return {
    simpleAction: jest.fn(),
  };
});

describe('FunctionalComponentWithHooks', () => {
  const mockClickAction = jest.fn();
  simpleAction.mockImplementation(mockClickAction);

  beforeEach(() => {
    getField.mockReturnValue('test');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it('should render default content', () => {
    const wrapper = renderShallow(FunctionalComponentWithHooks);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render new content', () => {
    getField.mockReturnValue('second');
    const wrapper = renderShallow(FunctionalComponentWithHooks);

    expect(wrapper).toMatchSnapshot();
  });

  it('should call click callback', () => {
    const wrapper = renderShallow(FunctionalComponentWithHooks);

    wrapper.simulate('click');
    expect(mockClickAction).toBeCalled();
  });
});
