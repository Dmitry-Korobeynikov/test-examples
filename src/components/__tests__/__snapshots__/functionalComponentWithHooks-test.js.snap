// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FunctionalComponentWithHooks should render default content 1`] = `
<div
  onClick={[Function]}
>
  test
</div>
`;

exports[`FunctionalComponentWithHooks should render new content 1`] = `
<div
  onClick={[Function]}
>
  second
</div>
`;
